llvm/lib/CodeGen/MachineScheduler.cpp

```diff
@@ -3979,6 +3990,17 @@ SUnit *GenericScheduler::pickNode(bool &IsTopNode) {
 
   LLVM_DEBUG(dbgs() << "Scheduling SU(" << SU->NodeNum << ") "
                     << *SU->getInstr());
+
+  if (IsTopNode) {
+    if (SU->NodeNum == TopIdx++)
+      ++NumInstrsInSourceOrderPreRA;
+  } else {
+    if (SU->NodeNum == BotIdx--)
```

Comment:
could you assert that BotIdx will always be less than NumRegionInstrs to prevent underflow?